File: app/Quotes/ocaml/src/core/backups.ml
(* Copyright 09-07-2018 ºDeme
   GNU General Public License - V3 <http://www.gnu.org/licenses/>
*)

let mk_date () = Date.(format "%Y%m%d" (now ()))

let mk_date2 () = Date.(
    let d = now () in
    Printf.sprintf "%s-%d" (format "%Y%m%d" d) (to_float d |> int_of_float)
  )

let clear_tmp () =
  let d = Path.((Cgi.home ()) ^ "tmp") in File.(
    del d;
    mkdir d
  )

let to_trash () =
  let home = Cgi.home () in
  Ext.zip
    Path.(home ^ "data")
    Path.(home ^ "trash" ^ (Printf.sprintf "%s.zip" (mk_date2 ())))

let unzip version_text =
  let adir = Cgi.home () in
  try (
    Ext.unzip Path.(adir ^ "tmp" ^ "back.zip") Path.(adir ^ "tmp");
    let source = Path.(adir ^ "tmp" ^ "data" ^ "version.txt") in (
      if File.exists source
      then if version_text = File.read_all source
        then ""
        else "restore:version is wrong"
      else "restore:version does not exist"
    )
  ) with e ->
    "restore:unzip"

let process version_text rq =
  match Cgi.rrq rq "rq" Json.rstring with
  | "backup" -> Cgi.(
      clear_tmp ();
      let nm = Printf.sprintf "%s%s.zip" (app_name ()) (mk_date ())
      and home = home () in (
        Ext.zip Path.(home ^ "data") Path.(home ^ "tmp" ^ nm);
        ok ["name", Json.String nm]
      )
    )
  | "restoreAbort" -> Cgi.(
      clear_tmp ();
      ok_empty ()
    )
  | "restoreAppend" -> Cgi.(
      let ch = File.aopen Path.((home ()) ^ "tmp" ^ "back.zip") in
      File.append_it_bin ch (
        rrq rq "data" Json.rstring |> B64.decode_bytes |> It.unary
      );
      close_out ch;
      ok_empty ()
    )
  | "restoreEnd" ->
    let fail = unzip version_text in (
      if fail <> "" then ()
      else (
        let home = Cgi.home () in
        let data = Path.(home ^ "data") in (
          to_trash ();
          File.del data;
          File.rename Path.(home ^ "tmp" ^ "data") data;
          clear_tmp ()
        )
      );
      Cgi.ok ["fail", Json.String fail]
    )
  | "restoreStart" -> Cgi.(
      clear_tmp ();
      let ch = File.wopen Path.((home ()) ^ "tmp" ^ "back.zip") in
      close_out ch;
      ok_empty ()
    )
  | "trash" ->
      let js = Path.((Cgi.home ()) ^ "trash") |> File.dir |> It.of_array |>
        Json.wit (fun s -> Json.String s)
      in
      Cgi.ok ["trash", js]
  | "backups" ->
      let js = Path.((Cgi.home ()) ^ "backups") |> File.dir |>
        It.of_array |> Json.wit (fun s -> Json.String s)
      in
      Cgi.ok ["backups", js]
  | "autorestore" -> Cgi.(
      let home = (home ()) in
      let source = Path.(home ^ "backups" ^ (rrq rq "file" Json.rstring)) in
      if File.exists source
      then (
        to_trash ();
        File.del Path.(home ^ "data");
        Ext.unzip source home;
        ok_empty ()
      )
      else raise (Failure (Printf.sprintf
        "Automatic backup '%s' not found"  source))
    )
  | "restoreTrash" -> Cgi.(
      let home = (home ()) in
      let source = Path.(home ^ "trash" ^ (rrq rq "file" Json.rstring)) in
      if File.exists source
      then (
        to_trash ();
        File.del Path.(home ^ "data");
        Ext.unzip source home;
        ok_empty ()
      )
      else raise (Failure (Printf.sprintf
        "Trash backup '%s' not found"  source))
    )
  | "clearTrash" ->
    let path = Path.(Cgi.(home ()) ^ "trash") in File.(
      del path;
      mkdir path;
      Cgi.ok_empty ()
    )
  | s -> raise (Failure (Printf.sprintf
    "Request '%s' is unknown in backups.rq" s))

File: app/Quotes/c/src/core/backups.c
// Copyright 05-Sept-2018 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "core/backups.h"
#include "dmc/ext.h"
#include "dmc/Date.h"
#include "dmc/b64.h"

static char *mk_date(void) {
  return date_format(date_now(), "%Y%m%d");
}

static char *mk_date2(void) {
	Date t = date_now();
  return str_printf("%s-%ld", date_format(t, "%Y%m%d"), t);
}

static Achar *read_trash(void) {
  Achar *fs = file_dir(path_cat(cgi_home(), "trash", NULL));
  Achar *rs = achar_new();
  EACH(fs, char, f) {
    achar_add(rs, path_name(f));
  }_EACH
	return rs;
}

static void to_trash(void) {
  char *date = mk_date2();
  char *source = path_cat(cgi_home(), "data", NULL);
  char *target = path_cat(
    cgi_home(), "trash", str_printf("%s.zip", date), NULL
  );
  ext_zip(source, target);
}

static void clear_tmp(void) {
  char *dir = path_cat(cgi_home(), "tmp", NULL);
  file_del(dir);
  file_mkdir(dir);
}

static char *unzip(char *app_name, char *data_version) {
  char *source = path_cat(cgi_home(), "tmp", "back.zip", NULL);
  char *target = path_cat(cgi_home(), "tmp", NULL);
  char *fail = "";
  TRY {
    ext_unzip(source, target);
  } CATCH(e) {
    fail = "restore:unzip";
  }_TRY
  if (*fail) {
    return fail;
  }

  source = path_cat(cgi_home(), "tmp", "data", "version.txt", NULL);
  if (!file_exists(source)) {
    return "restore:version does not exist";
  }
  char *good_version =str_printf(
    "%s\nData version: %s\n", app_name, data_version
  );
  char *version = file_read(source);
  if (!str_eq(version, good_version)) {
    return "restore:version is wrong";
  }
  return "";
}

//    ________________
CgiRp *backups_process(char *app_name, char* data_version, Mjson *mrq) {
//    TTTTTTTTTTTTTTTT
  char *home = cgi_home();
  char *rq = jmap_gstring(mrq, "rq");

  // -------------------------------------------------------------- backup
  if (str_eq(rq, "backup")) {
    clear_tmp();
    char *name = str_printf("%sBackup%s.zip", app_name, mk_date());

    ext_zip(
      path_cat(home, "data", NULL),
      path_cat(home, "tmp", name, NULL)
    );

    Mjson *m = mjson_new();
    jmap_pstring(m, "name", name);
    return cgi_ok(m);
  }

  // -------------------------------------------------_------ restoreStart
  if (str_eq(rq, "restoreStart")) {
    clear_tmp();
    LckFile *lck = file_wopen(path_cat(home, "tmp", "back.zip", NULL));
    file_close(lck);
    return cgi_ok(mjson_new());
  }

  // ------------------------------------------------------- restoreAppend
  if (str_eq(rq, "restoreAppend")) {
    LckFile *lck = file_aopen(path_cat(home, "tmp", "back.zip", NULL));
    file_write_bin(lck, b64_decode_bytes(jmap_gstring(mrq, "data")));
    file_close(lck);
    return cgi_ok(mjson_new());
  }

  // -------------------------------------------_------------ restoreAbort
  if (str_eq(rq, "restoreAbort")) {
    clear_tmp();
    return cgi_ok(mjson_new());
  }

  // ---------------------------------------------------------- restoreEnd
  if (str_eq(rq, "restoreEnd")) {
    char *fail = unzip(app_name, data_version);

    if (!*fail) {
      char *data = path_cat(home, "data", NULL);
      to_trash();
      file_del(data);
      file_rename(path_cat(home, "tmp", "data", NULL), data);
      clear_tmp();
    }

    Mjson *m = mjson_new();
    jmap_pstring(m, "fail", fail);
    return cgi_ok(m);
  }

  // --------------------------------------------------------- trashList
  if (str_eq(rq, "trashList")) {
    Mjson *m = mjson_new();
    jmap_parray(m, "list", (Arr *)read_trash(), (TO_JSON)json_wstring);
    return cgi_ok(m);
  }

  // ---------------------------------------------------------- clearTrash
  if (str_eq(rq, "clearTrash")) {
    char *path = path_cat(home, "trash", NULL);
    file_del(path);
    file_mkdir(path);
    return cgi_ok(mjson_new());
  }

  // -------------------------------------------------------- restoreTrash
  if (str_eq(rq, "restoreTrash")) {
    char *source = path_cat(home, "trash", jmap_gstring(mrq, "file"), NULL);

    if (!file_exists(source)) {
      return cgi_error(str_printf("Trash back '%s' not found", source));
    }

    to_trash();
    file_del(path_cat(home, "data", NULL));
    ext_unzip(source, home);

    return cgi_ok(mjson_new());
  }

  THROW("") "'%s': Unknown request in backups", rq _THROW
  // Unreacheable
  return NULL;
}
